lib/load-config.js

```diff
@@ -22,30 +22,131 @@ const CONFIG_FILE_NAMES = Object.freeze([
+
+    if (originalError instanceof Error) {
+      this.stack = originalError.stack
+    }
   }
+}
+
+const NO_EXPORTS = Symbol('no exports')
+
+function loadConfigFile (path, _di) {
+  const di = {
+    readFile: (path) => readFileSync(path, 'utf8'),
+    readJs: (path) => require(path),
+    ..._di
+  }
+
+  let config
+
+  const fileName = basename(path)
+  const ext = extname(path).toLowerCase()
+
+  if (EXTS_YAML_VALIDATOR.test(ext)) {
+    const JsYaml = require('js-yaml')
+    try {
+      // TODO: add YAML schema so that we get better errors for YAML users.
+      config = JsYaml.load(di.readFile(path))
+    } catch (error) {
+      if (typeof error === 'object') {
+        if (error instanceof JsYaml.YAMLException) {
+          throw new ConfigParsingError(
+            path,
+            'must contain a valid c8 configuration object.',
+            error
+          )
+        }
+      }
+
+      throw error
+    }
+
+    if (!config) {
+      // TODO: remove this check once we get YAML schema validation.
+      throw new ConfigParsingError(path, 'invalid configuration')
+    }
+  } else {
+    if (EXTS_JS_VALIDATOR.test(ext)) {
+      // Add a loader that allows us to check to see if anything was ever exported.
+      // Thank you to https://stackoverflow.com/a/70999950 for the inspiration. Please note that this code won't port to TypeScript nicely.
+      const extensions = module.constructor._extensions
+      const cjsLoader = extensions['.cjs']
+      const jsLoader = extensions['.js']
+      extensions['.cjs'] = extensions['.js'] = (module, filename) => {
+        module.exports[NO_EXPORTS] = filename
+        jsLoader(module, filename)
+      }
+      try {
+        config = di.readJs(path)
+      } finally {
+        // Undo the global state mutation, even if an error was thrown.
+        extensions['.cjs'] = cjsLoader
+        extensions['.js'] = jsLoader
+      }
+      if (NO_EXPORTS in config) {
+        throw new ConfigParsingError(
+          path,
+          'does not export a c8 configuration object.'
+        )
+      }
+    } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
+      try {
+        config = JSON.parse(di.readFile(path))
+      } catch (error) {
+        if (typeof error === 'object') {
+          if (error instanceof SyntaxError) {
+            throw new ConfigParsingError(
+              path,
+              'must contain a valid c8 configuration object.',
+              error
+            )
+          }
+        }
+
+        throw error
+      }
+    } else {
+      // If the user supplied a bad configuration file that we can't figure out how to read, then it's on them to solve it.
+      // Attempting to use some other config, even a default one, will result in unexpected behavior: aka ignoring the config that was explicitly specified is not intuitive.
+      throw new UnsupportedFileTypeError(
+        `Unsupported file type ${ext} while reading file ${path}`
+      )
+    }
```

Comment:
Refactor to remove an extra else statement.  Currently it reads:

```js
if (EXTS_YAML_VALIDATOR.test(ext)) {
...
} else {
   if (EXTS_JS_VALIDATOR.test(ext)) {
     ...
   } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
     ...
   } else {
     ...
   }
}

```

In the Review branch I changed this to:

```js
if (EXTS_YAML_VALIDATOR.test(ext)) {
  ...
} else if (EXTS_JS_VALIDATOR.test(ext)) {
  ...
} else if (ext === '.json' || fileName.slice(-2) === 'rc') {
  ...
} else {
  ...
}
```

Any questions or concerns?  

[Addressed](https://gist.github.com/mcknasty/80c8c814864241de05987dc3cae5fd54#file-parse-args-patch-diff-L123-L235) in the review branch.

Reply:
The structure of the if-else architecture was done specifically to support the correct injection of the relevant schema validators. Notice the TODO code comments for the expected placement of those validators. Notice that YAML would have a different validator than JSON and CJS as that will give a better DX than using a pure JSON schema validation after parsing the YAML.